### PR TITLE
ENG-18744 logging context mismatch and update snapshot coping

### PR DESCRIPTION
**Pull request by zeemanhe**
now snapshotting works better

### CODE REVIEW

File: src/ee/storage/persistenttable.cpp [hunk 1]
```diff
@@ -189,6 +191,16 @@ PersistentTable::~PersistentTable() {
...
 }
 
+void PersistentTable::checkContext(const std::string& operation) {
+   if ((isReplicatedTable() && !SynchronizedThreadLock::usingMpMemory()) ||
+      (!isReplicatedTable() && SynchronizedThreadLock::usingMpMemory())) {
+       std::stringstream message;
+       message << operation << (isReplicatedTable()?" REPLICATED ":" PARTITIONED ") << name().c_str()  << " with context " <<
+                 (SynchronizedThreadLock::usingMpMemory()?"REPLICATED":"PARTITIONED") << '\n';
+       string msg = message.str();
+       LogManager::getThreadLogger(LOGGERID_HOST)->log(voltdb::LOGLEVEL_ERROR, &msg);
```
Comment (kaikaizi) [inline src/ee/storage/persistenttable.cpp:201]: Also throw an exception after logging as error.

File: src/ee/storage/persistenttable.cpp [hunk 2]
```diff
@@ -189,6 +191,16 @@ PersistentTable::~PersistentTable() {
     }
 }
 
+void PersistentTable::checkContext(const std::string& operation) {
+   if ((isReplicatedTable() && !SynchronizedThreadLock::usingMpMemory()) ||
```
Comment (kaikaizi) [inline src/ee/storage/persistenttable.cpp:195]: check can be simplified as
```suggestion
   if (isReplicatedTable() != SynchronizedThreadLock::usingMpMemory()) {
```

File: src/ee/storage/persistenttable.cpp [hunk 3]
```diff
@@ -189,6 +191,16 @@ PersistentTable::~PersistentTable() {
     }
 }
 
+void PersistentTable::checkContext(const std::string& operation) {
```
Comment (kaikaizi) [inline src/ee/storage/persistenttable.cpp:194]: Minor: changing argument type to `char const*` can spare one object creation

File: src/ee/storage/persistenttable.cpp [hunk 4]
```diff
@@ -122,19 +122,20 @@ void PersistentTable::initializeWithColumns(TupleSchema* schema,
...
             TableTuple freshCopy(m_schema);
             freshCopy.move(fresh);
-
-            TableTuple original(m_schema);
-            original.move(const_cast<void*>(dest));
-            freshCopy.copyNonInlinedColumnObjects(original);
+            TableTuple source(m_schema);
+            source.move(const_cast<void*>(dest));
+            freshCopy.resetHeader();
+            freshCopy.copyForPersistentInsert(source);
```
Comment (kaikaizi) [inline src/ee/storage/persistenttable.cpp:138]: These 2 lines are the real fix, right?

File: src/ee/common/tabletuple.h [hunk 5]
```diff
@@ -469,7 +469,6 @@ class TableTuple {
     // required of the upcoming release or undo.
     void copyForPersistentUpdate(const TableTuple &source,
                                  std::vector<char*> &oldObjects, std::vector<char*> &newObjects);
-    void copyNonInlinedColumnObjects(TableTuple &target);
```
Comment (kaikaizi) [inline src/ee/common/tabletuple.h:472]: This method was introduced by our integration branch, right? Master branch does not have this method.

Comment (zeemanhe) [reply src/ee/common/tabletuple.h:472]: correct. no need anymore with the callback change

File: src/ee/storage/persistenttable.cpp (hunk 4 above)
Comment (zeemanhe) [reply src/ee/storage/persistenttable.cpp:138]: With the change, rejoining seems happy

File: src/ee/storage/persistenttable.cpp (hunk 3 above)
Comment (zeemanhe) [reply src/ee/storage/persistenttable.cpp:194]: will do

**Pull request merged**